Lib/concurrent/futures/_base.py

```diff
@@ -592,21 +595,38 @@ def map(self, fn, *iterables, timeout=None, chunksize=1):
         """
         if timeout is not None:
             end_time = timeout + time.monotonic()
+        if prefetch is None:
+            prefetch = self._max_workers
+        if prefetch < 0:
+            raise ValueError("prefetch count may not be negative")
 
-        fs = [self.submit(fn, *args) for args in zip(*iterables)]
+        argsiter = zip(*iterables)
+        initialargs = itertools.islice(argsiter, self._max_workers + prefetch)
+
+        fs = collections.deque(self.submit(fn, *args) for args in initialargs)
 
         # Yield must be hidden in closure so that the futures are submitted
         # before the first iterator value is required.
         def result_iterator():
+            nonlocal argsiter
             try:
-                # reverse to keep finishing order
-                fs.reverse()
                 while fs:
-                    # Careful not to keep a reference to the popped future
                     if timeout is None:
-                        yield fs.pop().result()
+                        res = fs.popleft().result()
                     else:
-                        yield fs.pop().result(end_time - time.monotonic())
```

Comment:
this doesn't cancel the currently waited on fut correctly:

```
import contextlib
import functools
import concurrent.futures
import threading
import sys


def fn(num, stop_event):
    if num == 1:
        stop_event.wait()
        return "done 1"

    if num == 2:
        return "done 2"


def main():
    stop_event = threading.Event()
    log = []

    with concurrent.futures.ThreadPoolExecutor(max_workers=1) as pool:

        def print_n_wait(ident):
            log.append(f"{ident=} started")
            try:
                stop_event.wait()
            finally:
                log.append(f"{ident=} stopped")

        fut = pool.submit(print_n_wait, ident="first")
        try:
            with contextlib.closing(pool.map(print_n_wait, ["second", "third"], timeout=1)) as gen:
                try:
                    next(gen)
                except concurrent.futures.TimeoutError:
                    print("timed out")
                else:
                    raise RuntimeError("timeout expected")
        finally:
            stop_event.set()

    assert log == ["ident='first' started", "ident='first' stopped"], f"{log=} is wrong"

if __name__ == "__main__":
    sys.exit(main())
    ```

Reply:
@graingert https://github.com/python/cpython/issues/95166 is resolved now, can you rebase this?